lrlex/src/lib/parser.rs

```diff
@@ -201,20 +200,22 @@ impl<StorageT: TryFrom<usize>> LexParser<StorageT> {
         if declaration_parameters.is_empty() {
             return Err(self.mk_error(LexErrorKind::UnknownDeclaration, off));
         }
-        let start_states: HashSet<(&str, Span)> = declaration_parameters
+        let start_states: Vec<(&str, Span)> = declaration_parameters
             .split_whitespace()
             .map(|name| {
                 let off = name.as_ptr() as usize - self.src.as_ptr() as usize;
                 let span = Span::new(off, off + name.len());
-                self.validate_start_state(span, name, errs)
+                (name, span)
             })
-            .filter_map(Result::transpose)
-            .collect::<LexInternalBuildResult<HashSet<(&str, Span)>>>()?;
+            .collect::<Vec<(&str, Span)>>();
 
         for (name, name_span) in start_states {
             let id = self.start_states.len();
-            let start_state = StartState::new(id, name, exclusive, name_span);
-            self.start_states.push(start_state);
+            if let Some(result) = self.validate_start_state(name_span, name, errs).transpose() {
+                let (name, name_span) = result?;
+                let start_state = StartState::new(id, name, exclusive, name_span);
+                self.start_states.push(start_state);
+            }
```

Comment:
Is it possible for there to be a `None` case here? I note it's silently ignored if so

Reply:
Yeah, that is the case where we encountered a duplicate, and the error was added to `errs` which is eventually caught in `parse`, so ignoring `None` is the right thing.

Reply:
One simplification here, could perhaps do is just return `Result<bool>` from `validate_start_state` since it is just returning its arguments?

Reply:
I went ahead and added a comment to this effect, and tried to simplify this branch in https://github.com/softdevteam/grmtools/pull/332/commits/661fc9340e80186728b974d90c35c18200493ee7.